org.eclipse.ltk.core.refactoring/src/org/eclipse/ltk/internal/core/refactoring/UndoableOperation2ChangeAdapter.java

```diff
@@ -348,7 +351,7 @@ private ExecuteResult executeChange(final IValidationCheckResultQuery query, IPr
 				monitor.done();
 			}
 		};
-		ResourcesPlugin.getWorkspace().run(runnable, pm);
+		ResourcesPlugin.getWorkspace().run(runnable, getSchedulingRule(), IWorkspace.AVOID_UPDATE, pm);
```

Comment:
It would be good to update commit message and explain why the change is needed, how this works and why this is safe to use extra lock in undo adapter.

I personally haven't yet tested this, but it would not surprise me if that would lead to similar deadlocks like the one seen in https://github.com/eclipse-platform/eclipse.platform.ui/issues/15

If any of the custom operations that are wrapped by this adapter do some non trivial locking, that might be dangerous. I also thinking about non-workspace rules that might be used - couldn't they lead to errors (rules cannot be nested?). 

So this need to be evaluated and if the change is safe, explained in commit why.

Reply:
Hint: updating commit message will automatically trigger rebuild.

Reply:
> So this need to be evaluated and if the change is safe, explained in commit why.

OK, new commit message is better but still doesn't explain **why** it is safe, just says it **is** safe.

The explanation that would simplify life of the reviewers could be:

The original call to `ResourcesPlugin.getWorkspace().run(runnable, pm);` is equivalent to the call of   `Workspace.run(IWorkspaceRunnable, IProgressMonitor)`, which is implemented internally as `run((ICoreRunnable) action, defaultRoot, IWorkspace.AVOID_UPDATE, monitor);` - means, changing the original code to proposed one is safe and doesn't change any behavior because the executed code it is identical at the end.